### ISSUE TITLE
[BUG]

### ISSUE THREAD

**Issue by kspradheep** [rancher/rancher#42371]
**Rancher Server Setup**
- Rancher version: v2.7.5
- Installation option (Docker install/Helm Chart): Helm Chart
   - If Helm Chart, Kubernetes Cluster and version (RKE1, RKE2, k3s, EKS, etc): EKS

**Information about the Cluster**
- Kubernetes version: 1.26

**User Information**
- What is the role of the user logged in? Admin

**Describe the bug**
Configured Rancher auth provider using Keycloak as per doc https://ranchermanager.docs.rancher.com/v2.7/how-to-guides/new-user-guides/authentication-permissions-and-global-configuration/authentication-config/configure-keycloak-oidc
Authentication using Keycloak works but failing to search user and groups principles due to Rancher keycloak provider client creation fails and unable to refresh tokens. Hence unable to add authorized groups or users.

Keycloak version: v21.1.2
Keycloak url: https://abc.com/auth/
Keycloak realm: sso

**To Reproduce**
1. Create rancher client in keycloak
2. Create users and assign groups in keycloak
3. Configure Rancher auth with Keycloak client. Authentication will be succeeded.
4. Select this option and search for authorized users or groups
![image](https://github.com/rancher/rancher/assets/68883779/c1c158ec-7367-4e58-8ec3-b1974a313aa8)

**Result**
2023/08/10 06:13:20 [INFO] Updating user user-gd8fc. Adding principal
2023/08/10 06:13:20 [ERROR] Error refreshing token principals, skipping: oauth2: cannot fetch token: 400 Bad Request
Response: {"error":"invalid_grant","error_description":"Invalid refresh token"}
2023/08/10 06:13:20 [ERROR] Error refreshing token principals, skipping: oauth2: cannot fetch token: 400 Bad Request
Response: {"error":"invalid_grant","error_description":"Invalid refresh token"}
2023/08/10 06:14:03 [ERROR] [keycloak oidc] getFromKeyCloakByID: GET request failed. url: https://abc.com/auth/admin/realms/sso/users/a892db5c-b773-422b-9501-070e89b79166, err: PermissionDenied 403: access denied
2023/08/10 06:14:03 [ERROR] error syncing 'user-gd8fc': handler mgmt-auth-userattributes-controller: PermissionDenied 403: access denied, requeuing
2023/08/10 06:15:02 [ERROR] [keycloak oidc] searchPrincipals: GET request failed. url: https://abc.com/auth/admin/realms/sso/users?search=kk, err: PermissionDenied 403: access denied
2023/08/10 06:15:02 [ERROR] [keycloak oidc] SearchPrincipals: problem searching keycloak: PermissionDenied 403: access denied
2023/08/10 06:16:03 [ERROR] [keycloak oidc] getFromKeyCloakByID: GET request failed. url: https://abc.com/auth/admin/realms/sso/users/a892db5c-b773-422b-9501-070e89b79166, err: PermissionDenied 403: access denied.

**Expected Result**
Rancher keycloak provider should create http client and send correct token request to succeed refresh token principals.

**Comment by tonowie1** [comment]
Rancher 2.7.5 installed using Helm (v1.25.11+rke2r1)

having the same issue I think. Admin has some permission issues around principals search. After Keycloak is enabled it is no longer possible to manage project access via UI because POST to /v3/principals?action=search fails with 403 and user/group lookup does not work. GET on /v3/principals?action=search returns groups and current user. This disables lookup even for the local users. Luckily this is sandbox env so I will be disabling Keycloak before solution is available

**Comment by kspradheep** [comment]
@tonowie1 what version of keycloak are you using? Did you have to do anything different than what listed in Rancher doc https://ranchermanager.docs.rancher.com/v2.7/how-to-guides/new-user-guides/authentication-permissions-and-global-configuration/authentication-config/configure-keycloak-oidc?

**Comment by anupama2501** [comment]
Hello @tonowie1 I'm on keycloak version 19.0.2. 
I tried the scenario you mentioned above and when the access is set to `Allow members of clusters and projects, plus authorized users & groups` , I was able to search for the keycloak users.

Could you make sure the session was active for the client in the keycloak server >> clients >> sessions?

Could you let us know if you have the following client scopes set on your keycloak users?
```
realm-management  >> client query-users /query-groups /view-users
```

**Comment by kspradheep** [comment]
Hi @anupama2501 after adding client scopes to keycloak users it worked. I can search users and groups now. Thanks for the help.

**Comment by tonowie1** [comment]
@kspradheep I do not have answer for you. Updated my edit to make it clear that rancher worked with keycloak disabled, it just took a bit of time. Keycloak version is 21 or 22. 
Based on what @anupama2501 posted it appears that Keycloak OIDC auth provider for Rancher uses not only OIDC but also keycloak REST API. I will have the suggestion tested out and post update here

**Comment by tonowie1** [comment]
my issue seems to be unrelated to this. Also getting 403 on principals endpoint but it is not resolved by the suggested changes. Will raise separate ticket

**Comment by tonowie1** [comment]
actually reviewing logs I have exactly the same issue... is there a way not to use the REST API of keycloak? I want just the OIDC auth provider

**Comment by anupama2501** [comment]
I apologize for the inconvenience. It appears that you're encountering an issue with an expired token. Have you tried logging out and then logging back in ? @eliaoggian

**Comment by eliaoggian** [comment]
Yes, I tried that already, I have this issue since months. It looks like the refresh token is never valid at all, as I can see the same errors even after a fresh rancher server restart. However, doing the requests manually with Postman works fine.

**Comment by tonowie1** [comment]
@anupama2501 would it be possible to update the documentation in order to clearly state what is the setup required to get Keycloak OIDC auth provider working? I think it would help a lot to avoid people running into these issues. Documentation mentions only OIDC setup but there is clearly some undocumented requirement to perhaps give some users permissions (I am not clear who needs which permissions) to access keycloak REST API

**Comment by eliaoggian** [comment]
@prachidamle No, the endpoints are exactly the same. It would be helpful to see exactly the request and parameters that is being sent to keycloak. I only can see it from the keycloak logs, but I cannot see the token that is being used to do the request.

I am on keycloak 1.18, I am now testing on a complete new instance of keycloak, and if it still doesn't work I'll try updating keycloak to a higher version. 

I agree with @tonowie1 , the documentation seems to be incomplete, and it is not clear if the requests againtst the keycloak APIs are done impersonating the currently logged in user (e.g myself) or if it uses a service account for the oidc client.

**Comment by eliaoggian** [comment]
I tested on a fresh installed keycloak server on version 18, with a new realm and a new client configured according to the instructions here: https://ranchermanager.docs.rancher.com/how-to-guides/new-user-guides/authentication-permissions-and-global-configuration/authentication-config/configure-keycloak-oidc 

Initially the queries were still not working. But adding the permissions suggested by @anupama2501 to the user fixed it. Therefore these roles are necessary on the users that will do the queries on the Rancher GUI:
```
Users > Role Mappings > Client Roles >  realm-management  
 + query-users
 + query-groups
 + view-users
```

 This needs to be documented and described.

I can also see from the logs of Rancher that the refresh token is working:
```
2023/09/07 13:40:14 [DEBUG] [generic oidc] RefeshAndUpdateToken: attempting to refresh access token
2023/09/07 13:40:14 [DEBUG] [generic oidc] UpdateToken: access token has been refreshed
2023/09/07 13:40:14 [DEBUG] [generic oidc] UpdateToken: saving refreshed access token
```

I'll compare the settings of the client I have on the other keycloak instance and try to undertstand what's causing the issue. 
I'll report my findings here.

Thanks for your help!

**Comment by eliaoggian** [comment]
I finally managed to get it working. 

The problem was that the requests to the keycloak API on the /auth/admin path were blocked from the Web server for security reasons. After fixing that and adding the realm-management client roles everything is working fine. 

I created a PR to add that section to the documentation, but it has not been reviewed yet.

**Comment by klippo** [comment]
We have kind of the same issue, but we don't have a 403 access denied error, but we can't renew the tokens.

Everything is configured as suggested by @eliaoggian in the PR.
```
rancher-65b546b454-qmkfw rancher 2023/10/05 09:02:33 [ERROR] [generic oidc] refetchGroupPrincipals: error getting user by principalID: can't find unique user for principal keycloakoidc_user://UUID

rancher-65b546b454-qmkfw rancher 2023/10/05 09:02:33 [ERROR] Error refreshing token principals, skipping: can't find unique user for principal keycloakoidc_user://UUID

rancher-65b546b454-qmkfw rancher 2023/10/05 09:02:33 [ERROR] [keycloak oidc] GetPrincipal: error creating new http client: oauth2: token expired and refresh token is not set

rancher-65b546b454-qmkfw rancher 2023/10/05 09:02:33 [ERROR] error syncing 'u-zd29x': handler mgmt-auth-userattributes-controller: oauth2: token expired and refresh token is not set, requeuing
```

Any ideas what we're missing?

**Comment by sass1997** [comment]
Seems to have solved the issue on our side as well Rancher 1.7.5 Keycloak 22.04

**Comment by sass1997** [comment]
Update: It seems that this stuff isn't working anymore. @eliaoggian I can see in your logs that you have added keycloak via generic oidc on rancher side. Any reasons for this?

**Comment by thomashoell** [comment]
Same issue. The corresponding Keycloak logs are:
`{"timestamp":"2024-03-13T10:21:56.054Z","sequence":8816,"loggerClassName":"org.jboss.logging.Logger","loggerName":"org.keycloak.events","level":"WARN","message":"type=\"REFRESH_TOKEN_ERROR\", realmId=\"7ea2f150-e8fa-48e6-91da-d9638a01fdb0\", clientId=\"rancher\", userId=\"null\", ipAddress=\"10.254.254.1\", error=\"invalid_token\", grant_type=\"refresh_token\", client_auth_method=\"client-secret\"","threadName":"executor-thread-2","threadId":144,"mdc":{},"ndc":"","hostName":"keycloak-0","processName":"QuarkusEntryPoint","processId":1}`

**Comment by CTPEKO3AT0P** [comment]
I've run into the similar issue:
[DEBUG] Triggering auth refresh on user-6rpw7
2024-04-15T11:50:16.880668314+03:00 2024/04/15 08:50:16 [TRACE] REST UPDATE apis/management.cattle.io/v3//userattributes/user-6rpw7
2024/04/15 08:50:16 [TRACE] REST GET api//v1/cattle-global-data/secrets/keycloakoidcconfig-clientsecret
2024/04/15 08:50:17 [ERROR] [keycloak oidc] searchPrincipals: GET request failed. url: https://keycloak-test.orange-ftgroup.ru/auth/admin/realms/OINIS/users?search=Ale, err: PermissionDenied 403: access denied
2024/04/15 08:50:17 [ERROR] [keycloak oidc] SearchPrincipals: problem searching keycloak: PermissionDenied 403: access denied

I am @Fresh RKE2 1.27 HA AirGap Install Rancher 2.8.3

**Comment by baptisterajaut** [comment]
Exact same issue. Keycloak works at first - assuming the user to login with has the realm_management roles, but afterwards, after one hour it wants to refresh the token and fails because it seems my tokens only last 30 minutes, not an hour. It would be nice to have a custom token refresh interval.
